### PR TITLE
indexer: fix unwrap after wrap bug in a checkpoint batch

### PR DESCRIPTION
**Pull request by gegaowp**
## Description 

title

## Test plan 

```
cargo nextest run objects/wrap_unwrap
PASS [   7.594s] sui-graphql-e2e-tests::tests run_test::stable/objects/wrap_unwrap.move
```

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

**Comment by lxfind** [comment]
Can you adda few e2e tests as well?

**Comment by gegaowp** [comment]
> a few e2e tests

@lxfind yes I would like to and need some help, I saw we have a couple of ingestion tests here https://github.com/MystenLabs/sui/blob/main/crates/sui-indexer/tests/ingestion_tests.rs#L78

and I plan to create transactions that wraps & unwraps objects respectively, I tried to search `sui-test-transaction-builder` but would not find one there
https://github.com/MystenLabs/sui/blob/main/crates/sui-test-transaction-builder/src/lib.rs

I saw that `authority_tests` implements `wrap_object` via calling `object_basics`, it looks like there is a related `object_basics.move` but I am a bit puzzled how they all work together.
https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/unit_tests/authority_tests.rs#L4465

any wisdom, did I over-complicate things?

**Comment by lxfind** [comment]
For these I wonder if you could just add snapshot tests in sui-graphql-e2e-tests for now.

File: crates/sui-indexer/src/store/pg_indexer_store.rs [hunk 1]
```diff
@@ -2187,38 +2145,49 @@ fn make_objects_history_to_commit(
...
+// For overlappings of mutations and deletions, only keep one with higher version.
+// This is necessary b/c after this step, DB commit will be done in parallel and not in order.
+fn retain_latest_indexed_objects(
     tx_object_changes: Vec<TransactionObjectChangesToCommit>,
-) -> Vec<StoredObjectSnapshot> {
-    let deleted_objects = tx_object_changes
-        .clone()
+) -> (Vec<IndexedObject>, Vec<IndexedDeletedObject>) {
+    // Only the last deleted / mutated object will be in the map,
+    // b/c tx_object_changes are in order and versions always increment,
```
Comment (bmwill) [inline crates/sui-indexer/src/store/pg_indexer_store.rs:2156]: This function talks about retaining the one with the highest version but it doesn't seem like there are any comparisons against the version of each object. I assume this is the reason why? Do we want to have some logic that does a check?

Comment (gegaowp) [reply crates/sui-indexer/src/store/pg_indexer_store.rs:2156]: yes exactly, I started with the comparison approach on the other similar fix and @lxfind suggested this which seems a bit cleaner https://github.com/MystenLabs/sui/pull/18991/files#r1723999314 

and sure, I can add some asserting logic here to make sure.

File: crates/sui-indexer/src/store/pg_indexer_store.rs [hunk 2]
```diff
@@ -2187,38 +2146,84 @@ fn make_objects_history_to_commit(
...
+        .flat_map(|change| {
+            change
+                .changed_objects
+                .into_iter()
+                .map(|m| Either::Left((m.object.id(), m)))
+                .chain(
+                    change
+                        .deleted_objects
+                        .into_iter()
+                        .map(|d| Either::Right((d.object_id, d))),
```
Comment (lxfind) [inline crates/sui-indexer/src/store/pg_indexer_store.rs:2169]: It doesn't seem necessary to get the ID here, you could always extract them in the `fold`.

File: crates/sui-indexer/src/store/pg_indexer_store.rs [hunk 3]
```diff
@@ -2187,38 +2146,84 @@ fn make_objects_history_to_commit(
...
-    deleted_objects
-        .into_values()
-        .chain(latest_objects.into_values().map(StoredObjectSnapshot::from))
-        .collect()
+                (mutations, deletions)
+            },
+        );
+    (
+        mutations.into_values().collect(),
+        deletions.into_values().collect(),
```
Comment (lxfind) [inline crates/sui-indexer/src/store/pg_indexer_store.rs:2227]: Probably simpler if we just turn them into the Stored format here?

Comment (gegaowp) [reply crates/sui-indexer/src/store/pg_indexer_store.rs:2227]: I kept them in indexed so that `objects` and `objects_snapshot` can share codes here

**Review by lxfind — APPROVED** [review]

**Comment by lxfind** [comment]
What are your thoughts on the plan to fix our live dataset (including the previous object related fixes)?

**Comment by gegaowp** [comment]
@xun my current plan is to prioritize the schema cleanup tasks https://linear.app/mysten-labs/project/clean-indexer-schema-75ccb3c4ddc4/issues and start a backfill with new schemas from genesis, hopefully with some recent changes that are supposed to improve backfill speed, the e2e backfill can finish within a week or so, we can potentially have data populated for multiple changes without creating backfill instances for each. Backup plan is to leverage `objects_history` and try to patch the affected tables, I do not have a concrete query in mind yet, thoughts?

**Pull request merged**